### PR TITLE
Symmetric backup

### PR DESCRIPTION
**Pull request by uhoreg**
Implements https://github.com/matrix-org/matrix-doc/pull/3270

### CODE REVIEW

**Comment by t3chguy** [comment]
You will need to rebase against latest develop as Dave landed some TSification

File: src/crypto/aes.ts [hunk 1]
```diff
@@ -261,3 +261,16 @@ export function decryptAES(data: IEncryptedPayload, key: Uint8Array, name: strin
...
+
+/** Calculate the MAC for checking the key.
+ *
+ * @param {Uint8Array} key the key to use
+ * @param {string} [iv] The initialization vector as a base64-encoded string.
+ *     If omitted, a random initialization vector will be created.
+ * @return {Promise<object>} An object that contains, `mac` and `iv` properties.
+ */
+export function calculateKeyCheck(key, iv = undefined) {
+    return encryptAES(ZERO_STR, key, "", iv);
```
Comment (t3chguy) [inline src/crypto/aes.ts:275]: now that this method lives here can not remove it from SecretStorage?

File: src/crypto/backup.ts [hunk 2]
```diff
@@ -654,8 +684,120 @@ export class Curve25519 implements BackupAlgorithm {
     }
 }
 
+function randomBytes(size: number): Uint8Array {
+    const crypto: {randomBytes: (number) => Uint8Array} | undefined = getCrypto() as any;
+    if (crypto) {
+        // nodejs version
+        return crypto.randomBytes(size);
+    }
+    if (typeof window !== "undefined" && window.crypto) {
```
Comment (t3chguy) [inline src/crypto/backup.ts:693]: could shorthand to `window?.crypto`

File: src/crypto/backup.ts [hunk 3]
```diff
@@ -654,8 +684,120 @@ export class Curve25519 implements BackupAlgorithm {
     }
 }
 
+function randomBytes(size: number): Uint8Array {
+    const crypto: {randomBytes: (number) => Uint8Array} | undefined = getCrypto() as any;
```
Comment (t3chguy) [inline src/crypto/backup.ts:688]:
```suggestion
    const crypto: {randomBytes: (n: number) => Uint8Array} | undefined = getCrypto() as any;
```

**Comment by uhoreg** [comment]
Argh!  I just merged in develop last night! :stuck_out_tongue:

File: src/crypto/backup.ts [hunk 4]
```diff
@@ -654,8 +684,120 @@ export class Curve25519 implements BackupAlgorithm {
...
+        // browser version
+        const buf = new Uint8Array(size);
+        window.crypto.getRandomValues(buf);
+        return buf;
+    }
+    throw new Error("No usable crypto implementation");
+}
+
+export class Aes256 implements BackupAlgorithm {
+    public static algorithmName = "org.matrix.msc3270.v1.aes-hmac-sha2";
```
Comment (t3chguy) [inline src/crypto/backup.ts:703]: can we use `UnstableValue` here?

File: src/crypto/backup.ts [hunk 5]
```diff
@@ -654,8 +684,120 @@ export class Curve25519 implements BackupAlgorithm {
...
+
+    constructor(
+        public authData: AuthData,
+        private key: Uint8Array,
```
Comment (t3chguy) [inline src/crypto/backup.ts:707]: readonly for sanity?

File: src/crypto/keybackup.ts [hunk 6]
```diff
@@ -35,11 +35,13 @@ export interface IKeyBackupRoomSessions {
...
-        signatures: ISignatures;
-        private_key_salt: string;
-        private_key_iterations: number;
+        public_key?: string;
+        signatures?: ISignatures;
+        private_key_salt?: string;
+        private_key_iterations?: number;
         private_key_bits?: number;
+        iv?: string;
+        mac?: string;
```
Comment (t3chguy) [inline src/crypto/keybackup.ts:44]: Can we not have a discriminated union or XOR here? As presumably there's two different payloads here and not a bunch of fields that can all occur at once.

File: src/crypto/key_passphrase.ts [hunk 7]
```diff
@@ -15,18 +15,13 @@ limitations under the License.
...
 const DEFAULT_BITSIZE = 256;
 
-/* eslint-disable camelcase */
-interface IAuthData {
-    private_key_salt: string;
-    private_key_iterations: number;
-    private_key_bits?: number;
-}
-/* eslint-enable camelcase */
+type IAuthData = IKeyBackupInfo["auth_data"];
```
Comment (t3chguy) [inline src/crypto/key_passphrase.ts:24]: Types shouldn't have the `I` prefix

**Review by t3chguy — APPROVED** [review]
:shipit:

**Pull request merged**